plugins/org.jkiss.dbeaver.registry/src/org/jkiss/dbeaver/registry/task/TaskManagerImpl.java

```diff
@@ -325,9 +328,33 @@ public DBTTaskRunStatus runTask(@NotNull DBRProgressMonitor monitor, @NotNull DB
     public TaskRunJob scheduleTask(@NotNull DBTTask task, @NotNull DBTTaskExecutionListener listener) {
         final TaskRunJob runJob = createJob((TaskImpl) task, listener);
         runJob.schedule();
+        if (serviceJob == null) {
+            serviceJob = createServiceJob();
+            serviceJob.setSystem(true);
+            serviceJob.schedule();
+        }
         return runJob;
     }
 
+    private Job createServiceJob() {
+        return new Job("Service canceling job") {
+            @Override
+            protected IStatus run(
+                IProgressMonitor monitor) {
+                for (TaskRunJob taskJob : runningTasks) {
+                    if (taskJob.isFinished() || taskJob.isCanceled()) {
+                        continue;
+                    }
+                    if (taskJob.checkCancelation()) {
+                        taskJob.setCancelByExecutionTime();
+                    }
```

Comment:
Why are you delegating checks and cancellations to the callee? Do everything within one call in the job itself.